Sarah UI integration with data layer

SarahPanther opened the pull request:
- MainActivity's recycler view now updates from the database (shows diary message and tag)
- DiaryActivity now updates existing inhaler usage events with DiaryEntry data (diary message and tag)
- Added DAO methods to update existing inhaler usage events with DiaryEntry data and WearableData
- Changed InhalerUsageEvent to be constructed with inner objects initialized (with placeholder values) so the DAO may update the existing InhalerUsageEvents without disturbing the other existing inner objects (updating DiaryEntry will not overwrite an existing WearableData, etc.)

Note: because of changes to the database, to run this app, you will need to uninstall the existing Breathe app from your phone as we have not created a migration strategy and Android will not let you install and run it.

SarahPanther committed:
Got rid of redundant dependency and changed android.arch.persistence dependencies to androidx

SarahPanther committed:
Added ViewModel and LiveData features - have not tested as tests need to be modified to accomadate the LiveData type from the BreahteDao

SarahPanther committed:
Added some additional test methods to the DAO so I could test without using the wrapper LiveData.  Updated the dependencies to include ViewModel and LiveData dependencies.  Updated main activity to pull values from database.  Currently, nothing in the RecyclerView is showing up (not even the null text --- for no IUE), so debugging that.  However, the app passes the modified tests and runs (just nothing shows up in the RecyclerView).  Added some Log statements to debug.

SarahPanther committed:
Made removed final from some variables to see if that would fix UI and migrated project to AndroidX

SarahPanther committed:
Revert "Made removed final from some variables to see if that would fix UI and migrated project to AndroidX"

This reverts commit 428dd660fd46654f49a42e27b394ef446481bfaa.

SarahPanther committed:
Changed populating database with fake IUEs from onCreate() to onOpen() which fixed the UI -> now updates from the database and shows the fake IUEs that were added when the database is created. Note: when the database is populated with these placeholder IUEs, all previously stored IUEs are deleted.  This should be removed once we can receive real data from the inhaler and are ready for users to actually use our system.

SarahPanther committed:
Progress bar now updates from database/LiveData

SarahPanther committed:
Changed BreatheRoomDatabase to only add placeholder IUEs if the database was empty and not to clear itself everytime it is opened (this is for testing purposes - final app should not add placeholder IUEs)

SarahPanther committed:
Created access to the BreatheDao deleteAll() method from the BreatheViewModel (for testing/app development purposes ONLY).  Will need to add a button to the UI later to call this ViewModel deleteAll() method.  This will hopefully allow us to test certain behavior more clearly as we continue to develop the app and integrate it with the inhaler.

SarahPanther committed:
Added InhalerUsageEvent timestamp to be passed on click via an intent between the main activity and the diary activity.

SarahPanther committed:
Diary Entry Activity now receives Instant timestamp string from Intent and sets header text to show the IUE's date

SarahPanther committed:
Exposed Dao update() Inhaler Usage Event through ViewModel (and repository).  Added feature that receives existing diary message at diary entry activity and allows user to update it.  Clicking the save button sends the message as a reply to main activity.

SarahPanther committed:
Main Activity saves a message to the database and passes it to the diary entry activity to be edited.  Still need to test this behavior and make sure nothing strange is happening. Considering moving the database saving part to the DiaryEntryActivity to make saving the tag easier/deal with less intents

SarahPanther committed:
Made a separate ViewModel instance for the progress bar (separate from recycler view)

SarahPanther committed:
Moved saving diary message to diary entry activity to avoid sending an intent back to the main activity

SarahPanther committed:
Diary timeline in main activity now shows message

SarahPanther committed:
DiaryEntryActivity saves Tag with the DiaryEntry into the InhalerUsageEvent in the database.

SarahPanther committed:
Added tag and some formatting to the diary entry timeline in the main activity

SarahPanther committed:
Changed database DAO to only update DiaryEntry and only update WearableData (stop clobbering existing data)

SarahPanther committed:
Actually add the files for the tag fix

SarahPanther committed:
Added comments and cleaned up code

SarahPanther committed:
Corrected some comments

SarahPanther commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     */
+    @Query("UPDATE InhalerUsageEvent_table SET tag = :tag, message = :diaryMessage " +
+            "WHERE Inhaler_Usage_Event_UTC_ISO_8601_date_time = :timeStamp")
+    int updateDiaryEntry(Instant timeStamp, Tag tag, String diaryMessage);
```
After talking to you about this in our group meeting, I think you are right. I could instead pass in and update the whole DiaryEntry object instead of initializing it with placeholder values when the InhalerUsageEvent is constructed.

I haven't tested doing this (passing the whole DiaryEntry object into the update function) yet, but I can when we refactor and see if it works the same.

SarahPanther commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+            "humidity = :humid, character = :character, " +
+            "digit = :digit " +
+            "WHERE Inhaler_Usage_Event_UTC_ISO_8601_date_time = :inhalerUsageTimeStamp")
+    int updateWearableData(Instant inhalerUsageTimeStamp, Instant wearableDataTimeStamp,
```
The same comment for updateDiaryEntry() goes for this method also.

Youssef-Beltagy commented on Breathe/.gitignore:
```diff
@@ -12,3 +12,4 @@
 /captures
 .externalNativeBuild
 .cxx
+.idea/dictionaries/panther.xml
```
Please put this in your personal exclude.

Youssef-Beltagy commented on Breathe/app/build.gradle:
```diff
@@ -54,12 +53,37 @@ dependencies {
+    def room_version = "2.2.6"
+
+    implementation "androidx.room:room-runtime:$room_version"
     annotationProcessor 'androidx.room:room-compiler:2.2.6'
```
use room_version here

Youssef-Beltagy commented on Breathe/app/build.gradle:
```diff
@@ -54,12 +53,37 @@ dependencies {
+
+    // Annotation processor
+    annotationProcessor "androidx.lifecycle:lifecycle-compiler:$lifecycle_version"
+    // alternately - if using Java8, use the following instead of lifecycle-compiler
```
Going forward, let's use Java8.

Youssef-Beltagy commented on Breathe/app/src/main/res/layout/iue_item.xml:
```diff
@@ -1,12 +1,33 @@
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
```
It is weird that height is not wrap_content.

Youssef-Beltagy commented on Breathe/app/src/main/res/layout/iue_item.xml:
```diff
@@ -1,12 +1,33 @@
+        android:orientation="vertical">
+
+        <TextView
+            android:id="@+id/timestamp_textview_in_recycler_main"
```
It is somewhat annoying that the name is so long and textview is not at the end. But this is not important.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     * method to work correctly.
+     *
+     * @return all InhalerUsageEvents from the InhalerUsageEvent_table in lexicographical,
+     * descending order of string timestamp (because of UTC format, this is in chronological order)
```
It is because of the iso 8061 format. No need to fix this now.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+
 
-    // Update one or more InhalerUsageEvents
+    // TODO: maybe we should never use this because it overwites/"clobbers" our existing IUEs
```
It is fine to keep it as long as we explicitly write it overwrites in the comments/documentation.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+
+    /**
+     * Note: Use this one to update an existing inhalerUsageEvent with DiaryEntry data so the
+     * existing other inner objects (WearableData, WeatherData) don't get "clobbered"
```
I don't know what you mean by clobbered. Do you mean overwritten?

SarahPanther replied:
Yes

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     * @param diaryMessage
+     * @return the number of records updated (should only be 1)
+     */
+    @Query("UPDATE InhalerUsageEvent_table SET tag = :tag, message = :diaryMessage " +
```
This interface is weird. Why didn't you simply have it take an IUE? Now, you are taking away the abstraction of an IUE.

SarahPanther replied:
Because if we don't update simply the diary entry object, the other inner objects get overwritten. We may pass in a DiaryEntry object (and that is probably more elegant), but if we pass in an IUE, we will need to retrieve the single specific IUE that is being edited.  This takes a long time since the call happens "asynchronously" but needs to happen synchronously when the user saves the diary entry.

I can talk to you more about this during the meeting but this is the whole issue I was telling you a couple weeks ago about that this method solves.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     * @param digit
+     * @return the number of records updated (should only be 1)
+     */
+    @Query("UPDATE InhalerUsageEvent_table " +
```
As above, this interface looks weird. It also means whoever changes the data of the smart pin will have to come here and update the Dao.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+                           float temp, float humid, char character, char digit);
+
+    /**
+     * Note: This method alone is fine but there is no wrapper method for this in the Breathe
```
This is true of all the DB select queries. I don't think there is anything we can do about it.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     */
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
+            "WHERE Inhaler_Usage_Event_UTC_ISO_8601_date_time == :timeStamp")
+    List<InhalerUsageEvent> getInhalerUsageEventWithTimeStamp(Instant timeStamp);
```
Out of curiosity, must it always return a List? Can't it just return a single IUE or a generic collection of IUE?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+     * @param secondDate - Date in range that occurred last
+     * @return InhalerUsageEvents between the parameter dates (inclusive) wrapped as LiveData
+     */
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
```
Please order by DESC

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+    LiveData<List<InhalerUsageEvent>> loadAllInhalerUsageEventsBetweenDates(Instant firstDate,
+                                                                            Instant secondDate);
+
+    // Methods for testing ONLY --------------------------------------------------------------------
```
I don't understand how this is any different from using the other getAllIUE.

SarahPanther replied:
One returns LiveData and one is a plain old list.  It is difficult to test with the wrapper LiveData() so I made one that just returns a list.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -19,27 +20,120 @@
+    List<InhalerUsageEvent> getAllIUEsTest();
 
-    // Return InhalerUsageEvents between the parameter dates (inclusive)
     @Query("SELECT * FROM InhalerUsageEvent_table " +
```
Same. Maintaining two things instead of just one is just a headache.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRepository.java:
```diff
@@ -11,17 +16,84 @@
  */
 public class BreatheRepository {
     private final BreatheDao breatheDao;
+    private final LiveData<List<InhalerUsageEvent>> allInhalerUsageEvents; // TODO: IDE wants to make this final, does this cause problems?
```
Making this final is not a problem. Just don't reassign it with " = new LiveData<List<InhalerUsageEvent>>()". Java objects are references, so the final is the reference. Basically, you are saying it will always point to the same location in memory. But you can still use add, erase, and clear with the list.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -1,5 +1,6 @@
 package com.ybeltagy.breathe;
```
Going forward, it makes more sense if you put the DB functionality in its own package, make the Dao methods protected, and only make the repository methods public.

Youssef-Beltagy replied:
The Dao methods must be public. Otherwise, I implemented everything else.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRepository.java:
```diff
@@ -11,17 +16,84 @@
+    }
+
+    /**
+     * TODO: maybe we should never use this because it "clobbers" our existing IUEs (unless we
```
Same comment as in the Dao.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRepository.java:
```diff
@@ -11,17 +16,84 @@
+     * @param tag
+     * @param diaryMessage
+     */
+    public void updateDiaryEntry(Instant timeStamp, Tag tag, String diaryMessage) {
```
Again, this interface is reallyyyy weird. Please use an IUE as a paramter and consider changing the method name as required.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRepository.java:
```diff
@@ -11,17 +16,84 @@
+     * @param character
+     * @param digit
+     */
+    public void updateWearableData(Instant inhalerUsageTimeStamp, Instant wearableDataTimeStamp, float temperature, float humidity, char character, char digit) {
```
Same as above. Going forward, it will be nice if we don't store a wearableData timestamp in the IUE table but only do that in the  WearableData table.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRepository.java:
```diff
@@ -11,17 +16,84 @@
+    // Methods for TESTING PURPOSES ONLY------------------------------------------------------------
+
+    /**
+     * IMPORTANT : This is for testing purposes ONLY -> to clear away placeholder IUEs we've created
```
Please remove that warning. If we will make a settings page, we can add a clear IUE data button there.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java:
```diff
@@ -35,11 +42,44 @@ public static BreatheRoomDatabase getDatabase(final Context context) {
+                if (breatheDao.getAnySingleInhalerUsageEvent().size() < 1) {
+                    // Check if there are any placeholder IUEs in the database
+                    // If there are none, make some placeholder IUEs and add them to the database
+                    Instant now = Instant.now();
```
You are inserting in ascending order when our use case is better portrayed by inserting in descending order.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java:
```diff
@@ -35,11 +42,44 @@ public static BreatheRoomDatabase getDatabase(final Context context) {
                     // create database
                     INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                             BreatheRoomDatabase.class, "Breathe_database")
+                            .addCallback(roomDatabaseCallback)
```
I wonder if there is something like that for the repository methods. Please explain clearly this is just for testing and add todo/fixme to delete it before we wrap up the app. If we will have a callback in the final deliverable, consider making a separate class in a separate file for the callback rather than making a private data member.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheViewModel.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public class BreatheViewModel extends AndroidViewModel {
+    private final BreatheRepository breatheRepository;
+    private final LiveData<List<InhalerUsageEvent>> allInhalerUsageEvents;
```
I'm somewhat confused why the Repository needs this list if you have it here anyway.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheViewModel.java:
```diff
@@ -0,0 +1,67 @@
+        breatheRepository.updateDiaryEntry(timeStamp,tag, diaryMessage);
+    }
+
+    public void updateWearableData(Instant inhalerUsageTimestamp, Instant wearableDataTimestamp,
```
There are so many layers of abstraction and the interface is weird. You can easily flip temperature and humidity.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheViewModel.java:
```diff
@@ -0,0 +1,67 @@
+        return allInhalerUsageEvents;
+    }
+
+    public void insert(InhalerUsageEvent inhalerUsageEvent) {
```
I'm somewhat confused why you have to add these methods here too?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -20,7 +25,20 @@
+    public static final int UPDATE_INHALER_USAGE_EVENT_REQUEST_CODE = 1;
+
+    // Intent extra for Diary Entry Activity
+    // TODO: extract these to string resources?
```
No, not to a string resource but to a centralized java class or something.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/Converters.java:
```diff
@@ -40,27 +35,43 @@ public static String toTimeStampString(Instant timeStamp) {
-    public static String toTagString(Tag tag){
-        return tag == null ? null : tag.toString();
+    public static int toTagIntValue(Tag tag) {
+        return tag != null ? tag.ordinal() : Tag.NULL.ordinal();
```
It might make more sense to make Null the default value in the constructor instead.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/Converters.java:
```diff
@@ -14,24 +14,19 @@
 
```
Why did you bother switching to ints?

Youssef-Beltagy replied:
This is already done. I don't know. I feel a little uncomfortable with having multiple points in the code that do something similar.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/DiaryEntryActivity.java:
```diff
@@ -1,20 +1,119 @@
+     *
+     * @param timeStampString
+     */
+    public void setHeaderDateForInhalerUsageEvent(String timeStampString) {
```
private

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/DiaryEntryActivity.java:
```diff
@@ -1,20 +1,119 @@
+     * @param messageEditText
+     */
+    public void setExistingDiaryMessage(String diaryEntryMessage, EditText messageEditText) {
+        if (!diaryEntryMessage.isEmpty()) {
```
check for null even if it should never be there because the app will crash for a null.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/DiaryEntryActivity.java:
```diff
@@ -1,20 +1,119 @@
+
+        // When the user presses the Save button, create a new Intent for the reply.
+        // The reply Intent will be sent back to the calling activity (in this case, MainActivity).
+        saveButton.setOnClickListener(view -> {
```
Simply make a method. Why are you attaching a lambda function?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/DiaryEntryActivity.java:
```diff
@@ -1,20 +1,119 @@
+            if (this.tag != null) {
+                tag = this.tag;
+            }
+            breatheViewModel.updateDiaryEntry(Instant.parse(timeStampString), tag, messageEditText.getText().toString());
```
So the default constructor is never called. And what handles null cases is actually the converter. Okay, good to know.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -29,22 +47,15 @@ protected void onCreate(Bundle savedInstanceState) {
-        }
-        renderDiaryView(eventList);
+        //  or put that in the resources folder if we will continue using it.
+        int totalDosesInCanister = 200;
```
We might want to discuss with clients whether they want to pass this in any different way rather than hard code it.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
+        medicineStatusBar.setMax(totalDosesInCanister);
+
+        // set doses taken shown in the ProgressBar
+        progressBarViewModel = new ViewModelProvider(this,
```
It is more appropriate to put this in the onCreate method.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
+
         // make adapter and provide data to be displayed
-        IUEListAdapter iueListAdapter = new IUEListAdapter(this, eventList);
+        IUEListAdapter iueListAdapter = new IUEListAdapter(this);
```
It might actually make more sense to pass the list as a parameter. https://developer.android.com/training/dependency-injection

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
         // set layout manager for recyclerView
         iueRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+
+        // Note: constructor in codelab did not work; searched for a long time and this fixed it:
```
Thank you.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
+                .get(BreatheViewModel.class);
+
+        // Updated cached copy of InhalerUsageEvents
+        diaryTimelineViewModel.getAllInhalerUsageEvents().observe(this, inhalerUsageEvents1 -> {
```
Does the IUE list reference inside the list adabter automatically update as well?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    public void launchDiaryEntryActivity(InhalerUsageEvent inhalerUsageEvent) {
```
private

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/MainActivity.java:
```diff
@@ -53,28 +64,85 @@ protected void onDestroy() {
+                inhalerUsageEvent.getInhalerUsageEventTimeStamp().toString());
+
+        // add existing message
+        String existingMessage = inhalerUsageEvent.getDiaryEntry() != null ?
```
In a sense, wouldn't it be better to have the Diary entry activity do all of this retrieving?

Youssef-Beltagy committed:
added updateWeatherData in Dao

- Refactored
- Deleted code that will never be used
- Moved classes into packages

SarahPanther committed:
Merge pull request #27 from Youssef-Beltagy/UI+data_integration_review

refactored, organized, and trimmed down
